diff -urN linux-6.12-rc6/drivers/gpu/drm/sun4i/sun8i_mixer.c linux-6.12-rc6-wip/drivers/gpu/drm/sun4i/sun8i_mixer.c
--- linux-6.12-rc6/drivers/gpu/drm/sun4i/sun8i_mixer.c	2024-10-27 23:52:02.000000000 +0100
+++ linux-6.12-rc6-wip/drivers/gpu/drm/sun4i/sun8i_mixer.c	2024-11-05 08:57:58.000000000 +0100
@@ -705,6 +705,15 @@
 	.vi_num		= 1,
 };
 
+static const struct sun8i_mixer_cfg sun8i_v853_mixer_cfg = {
+	.ccsc		= CCSC_MIXER0_LAYOUT,
+	.mod_rate	= 297000000,
+	.scaler_mask	= 0xf,
+	.scanline_yuv	= 2048,
+	.ui_num		= 1,
+	.vi_num		= 1,
+};
+
 static const struct sun8i_mixer_cfg sun50i_a64_mixer1_cfg = {
 	.ccsc		= CCSC_MIXER1_LAYOUT,
 	.mod_rate	= 297000000,
@@ -750,6 +759,10 @@
 		.data = &sun8i_v3s_mixer_cfg,
 	},
 	{
+		.compatible = "allwinner,sun8i-v853-de2-mixer",
+		.data = &sun8i_v853_mixer_cfg,
+	},
+	{
 		.compatible = "allwinner,sun20i-d1-de2-mixer-0",
 		.data = &sun20i_d1_mixer0_cfg,
 	},
