diff -urN linux-6.12-rc6/drivers/mtd/nand/spi/foresee.c linux-6.12-rc6-wip/drivers/mtd/nand/spi/foresee.c
--- linux-6.12-rc6/drivers/mtd/nand/spi/foresee.c	2024-10-27 23:52:02.000000000 +0100
+++ linux-6.12-rc6-wip/drivers/mtd/nand/spi/foresee.c	2024-10-30 11:38:30.000000000 +0100
@@ -81,6 +81,16 @@
 		     SPINAND_HAS_QE_BIT,
 		     SPINAND_ECCINFO(&f35sqa002g_ooblayout,
 				     f35sqa002g_ecc_get_status)),
+	SPINAND_INFO("F35SQA001G",
+		     SPINAND_ID(SPINAND_READID_METHOD_OPCODE_DUMMY, 0x71, 0x71),
+		     NAND_MEMORG(1, 2048, 64, 64, 1024, 20, 1, 1, 1),
+		     NAND_ECCREQ(1, 512),
+		     SPINAND_INFO_OP_VARIANTS(&read_cache_variants,
+					      &write_cache_variants,
+					      &update_cache_variants),
+		     SPINAND_HAS_QE_BIT,
+		     SPINAND_ECCINFO(&f35sqa002g_ooblayout,
+				     f35sqa002g_ecc_get_status)),
 };
 
 static const struct spinand_manufacturer_ops foresee_spinand_manuf_ops = {
