diff -urN linux-6.12-rc6/drivers/rtc/rtc-sun6i.c linux-6.12-rc6-wip/drivers/rtc/rtc-sun6i.c
--- linux-6.12-rc6/drivers/rtc/rtc-sun6i.c	2024-10-27 23:52:02.000000000 +0100
+++ linux-6.12-rc6-wip/drivers/rtc/rtc-sun6i.c	2024-11-07 14:50:16.000000000 +0100
@@ -367,6 +367,20 @@
 CLK_OF_DECLARE_DRIVER(sun50i_h5_rtc_clk, "allwinner,sun50i-h5-rtc",
 		      sun8i_h3_rtc_clk_init);
 
+static const struct sun6i_rtc_clk_data sun8i_v853_rtc_data = {
+	.rc_osc_rate = 16000000,
+	.fixed_prescaler = 32,
+	.has_prescaler = 1,
+	.has_out_clk = 1,
+};
+
+static void __init sun8i_v853_rtc_clk_init(struct device_node *node)
+{
+	sun6i_rtc_clk_init(node, &sun8i_v853_rtc_data);
+}
+CLK_OF_DECLARE_DRIVER(sun8i_v853_rtc_clk, "allwinner,sun8i-v853-rtc",
+		      sun8i_v853_rtc_clk_init);
+
 static const struct sun6i_rtc_clk_data sun50i_h6_rtc_data = {
 	.rc_osc_rate = 16000000,
 	.fixed_prescaler = 32,
@@ -863,6 +877,8 @@
 	{ .compatible = "allwinner,sun8i-r40-rtc" },
 	{ .compatible = "allwinner,sun8i-v3-rtc" },
 	{ .compatible = "allwinner,sun50i-h5-rtc" },
+	{ .compatible = "allwinner,sun8i-v853-rtc",
+		.data = (void *)RTC_LINEAR_DAY },
 	{ .compatible = "allwinner,sun50i-h6-rtc" },
 	{ .compatible = "allwinner,sun50i-h616-rtc",
 		.data = (void *)RTC_LINEAR_DAY },
