diff -urN linux-6.12-rc6/arch/arm/boot/dts/allwinner/sun8i-v851s-lizard.dts linux-6.12-rc6-wip/arch/arm/boot/dts/allwinner/sun8i-v851s-lizard.dts
--- linux-6.12-rc6/arch/arm/boot/dts/allwinner/sun8i-v851s-lizard.dts	1970-01-01 01:00:00.000000000 +0100
+++ linux-6.12-rc6-wip/arch/arm/boot/dts/allwinner/sun8i-v851s-lizard.dts	2024-11-05 13:31:21.843720982 +0100
@@ -0,0 +1,365 @@
+// SPDX-License-Identifier: (GPL-2.0+ or MIT)
+// Copyright (C) 2024 Andras Szemzo.andras@gmail.com>
+
+/dts-v1/;
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/leds/common.h>
+#include "sun8i-v851s.dtsi"
+
+/ {
+	model = "Yuzuki-lizard";
+	compatible = "allwinner,sun8i-v853", "allwinner,sun8i";
+
+	aliases {
+		serial0 = &uart0;
+		serial2 = &uart2;
+	};
+
+	chosen {
+		stdout-path = "serial2:115200n8";
+	};
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		riscv_reserved: riscv_dram {
+			compatible = "shared-dma-pool";
+			reg = <0x43c00000 0x00180000>;
+			no-map;
+		};
+
+		vdev0buffer: vdev0buffer@43d80000 {
+			/* 256k reserved for shared mem pool */
+			compatible = "shared-dma-pool";
+			reg = <0x43d80000 0x40000>;
+			no-map;
+		};
+
+		vdev0vring0: vdev0vring0@43dc0000 {
+			compatible = "shared-dma-pool";
+			reg = <0x43dc0000 0x20000>;
+			no-map;
+		};
+
+		vdev0vring1: vdev0vring1@43de0000 {
+			compatible = "shared-dma-pool";
+			reg = <0x43de0000 0x20000>;
+			no-map;
+		};
+	};
+
+	reg_vcc: vcc {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+	};
+
+	reg_vcc_1v8: vcc-1v8 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc-1v8";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+	};
+
+	reg_usbvbus: usbvbus {
+		compatible = "regulator-fixed";
+		regulator-name = "usbvbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		vin-supply = <&reg_vcc>;
+	};
+
+	reg_vcc_3v3: vcc-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		vin-supply = <&reg_vcc>;
+	};
+
+	reg_vdd_cpu: vdd-cpu {
+		compatible = "regulator-fixed";
+		regulator-name = "vdd-cpu";
+		regulator-min-microvolt = <900000>;
+		regulator-max-microvolt = <900000>;
+		vin-supply = <&reg_vcc>;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		status_led {
+			linux,default-trigger = "heartbeat";
+			label = "led:red";
+			gpios = <&pio 5 6 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	wifi_pwrseq: wifi-pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&pio 4 6 GPIO_ACTIVE_LOW>; /* PE6 */
+	};
+
+	dmic_codec: dmic-codec {
+		compatible = "dmic-codec";
+		num-channels = <2>;
+		#sound-dai-cells = <0>;
+	};
+
+	dmic-sound {
+		compatible = "simple-audio-card";
+		simple-audio-card,name = "DMIC";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		simple-audio-card,dai-link@0 {
+			reg = <0>;
+			format = "pdm";
+			frame-master = <&link0_cpu>;
+			bitclock-master = <&link0_cpu>;
+
+			link0_cpu: cpu {
+				sound-dai = <&dmic>;
+			};
+
+			link0_codec: codec {
+				sound-dai = <&dmic_codec>;
+			};
+		};
+	};
+};
+
+&osc24M {
+	clock-frequency = <24000000>;
+};
+
+&cpu0 {
+	cpu-supply = <&reg_vdd_cpu>;
+};
+
+&pio {
+	vcc-pa-supply = <&reg_vcc_3v3>;
+	vcc-pc-supply = <&reg_vcc_3v3>;
+	vcc-pd-supply = <&reg_vcc_3v3>;
+	vcc-pe-supply = <&reg_vcc_3v3>;
+	vcc-pf-supply = <&reg_vcc_3v3>;
+	vcc-ph-supply = <&reg_vcc_3v3>;
+
+	/omit-if-no-ref/
+	pwm_ph11_pin: pwm-ph11-pin {
+		pins = "PH11";
+		function = "pwm4";
+	};
+};
+
+&uart0 {
+	pinctrl-0 = <&uart0_ph9_pins>;
+	pinctrl-names = "default";
+	status = "disabled";
+};
+
+&uart2 {
+	pinctrl-0 = <&uart2_pe12_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&wdt {
+	status = "okay";
+};
+
+&spi0 {
+	pinctrl-0 = <&spi0_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+
+	flash@0 {
+		compatible = "spi-nand";
+		spi-max-frequency = <75000000>;
+		reg = <0>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "awboot";
+				reg = <0x00000000 0x00020000>;	/* 128K */
+			};
+
+			partition@20000 {
+				label = "dtb";
+				reg = <0x00020000 0x00020000>;  /* 128k */
+			};
+
+			partition@40000 {
+				label = "kernel";
+				reg = <0x00040000 0x00800000>;  /* 8MB */
+			};
+
+			partition@840000 {
+				label = "rootfs";
+				reg = <0x00840000 0x077c0000>;
+			};
+		};
+	};
+};
+
+&spi3 {
+	pinctrl-0 = <&spi3_pins>;
+	pinctrl-names = "default";
+	status = "disabled";
+
+	ice40: fpga@0 {
+	    compatible = "lattice,ice40-fpga-mgr";
+	    reg = <0>;
+	    spi-max-frequency = <10000000>;
+	    cdone-gpios = <&pio 4 10 GPIO_ACTIVE_HIGH>; /* PE10 */
+	    reset-gpios = <&pio 4 11 GPIO_ACTIVE_LOW>;  /* PE11 */
+	};
+};
+
+&i2c2 {
+	pinctrl-0 = <&i2c2_ph11_pins>;
+	pinctrl-names = "default";
+	status = "disabled";
+};
+
+&pwm {
+	pinctrl-0 = <&pwm_ph11_pin>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&mmc0 {
+	bus-width = <4>;
+	broken-cd;
+	disable-wp;
+	vmmc-supply = <&reg_vcc_3v3>;
+	vqmmc-supply = <&reg_vcc_3v3>;
+	pinctrl-0 = <&mmc0_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&mmc1 {
+	bus-width = <4>;
+	mmc-pwrseq = <&wifi_pwrseq>;
+	non-removable;
+	vmmc-supply = <&reg_vcc_3v3>;
+	vqmmc-supply = <&reg_vcc_3v3>;
+	pinctrl-0 = <&mmc1_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+
+	xr829: wifi@1 {
+		reg = <1>;
+		interrupt-parent = <&pio>;
+		interrupts = <4 7 IRQ_TYPE_LEVEL_HIGH>; /* PE7 */
+		interrupt-names = "host-wake";
+	};
+};
+
+&usb_otg {
+	dr_mode = "otg";
+	status = "okay";
+};
+
+&ehci0 {
+	status = "okay";
+};
+
+&ohci0 {
+	status = "okay";
+};
+
+&usbphy {
+	usb0_vbus-supply = <&reg_usbvbus>;
+	usb1_vbus-supply = <&reg_usbvbus>;
+	usb0_id_det-gpio = <&pio 4 9 GPIO_ACTIVE_LOW>;
+	status = "okay";
+};
+
+&gpadc {
+	status = "okay";
+
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	channel@0 {
+	    reg = <0>;
+	};
+};
+
+&de {
+	status = "okay";
+};
+
+&dphy {
+	status = "okay";
+};
+
+&tcon_lcd0 {
+	status = "okay";
+};
+
+&dsi {
+	status = "okay";
+
+	pinctrl-0 = <&dsi_2lane_pins>;
+	pinctrl-names = "default";
+
+	ports {
+	    #address-cells = <1>;
+	    #size-cells = <0>;
+
+	    dsi_out: port@0 {
+		reg = <0>;
+
+		dsi_out_panel: endpoint {
+		    remote-endpoint = <&panel_out_dsi>;
+		};
+	    };
+	};
+
+	panel@0 {
+	    compatible = "lg,lh500wx1-sd03";
+	    reg = <0>;
+
+	    port {
+		panel_out_dsi: endpoint {
+		    remote-endpoint = <&dsi_out_panel>;
+		};
+	    };
+	};
+};
+
+&npu {
+	status = "okay";
+};
+
+&audio_codec {
+        allwinner,audio-routing = "Line Out", "LINEOUT",
+                                  "Speaker", "LINEOUT";
+	status = "okay";
+};
+
+&dmic {
+	pinctrl-0 = <&dmic_ph9_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&msgbox {
+	status = "okay";
+};
+
+&remoteproc {
+	clock-frequency = <600000000>;
+	firmware-name = "v851s-riscv-firmware.elf";
+	status = "okay";
+};
