diff -urN linux-6.7/drivers/pmdomain/sunxi/sun20i-ppu.c linux-6.7-saam/drivers/pmdomain/sunxi/sun20i-ppu.c
--- linux-6.7/drivers/pmdomain/sunxi/sun20i-ppu.c	2024-01-07 21:18:38.000000000 +0100
+++ linux-6.7-saam/drivers/pmdomain/sunxi/sun20i-ppu.c	2024-01-11 09:12:08.900246815 +0100
@@ -182,11 +182,26 @@
 	.num_domains	= ARRAY_SIZE(sun20i_d1_ppu_pd_names),
 };
 
+static const char *const sun8i_v853_ppu_pd_names[] = {
+	"RISCV",
+	"NPU",
+	"VE",
+};
+
+static const struct sun20i_ppu_desc sun8i_v853_ppu_desc = {
+	.names		= sun8i_v853_ppu_pd_names,
+	.num_domains	= ARRAY_SIZE(sun8i_v853_ppu_pd_names),
+};
+
 static const struct of_device_id sun20i_ppu_of_match[] = {
 	{
 		.compatible	= "allwinner,sun20i-d1-ppu",
 		.data		= &sun20i_d1_ppu_desc,
 	},
+	{
+		.compatible	= "allwinner,sun8i-v853-ppu",
+		.data		= &sun8i_v853_ppu_desc,
+	},
 	{ }
 };
 MODULE_DEVICE_TABLE(of, sun20i_ppu_of_match);
diff -urN linux-6.7/include/dt-bindings/power/allwinner,sun8i-v853-ppu.h linux-6.7-saam/include/dt-bindings/power/allwinner,sun8i-v853-ppu.h
--- linux-6.7/include/dt-bindings/power/allwinner,sun8i-v853-ppu.h	1970-01-01 01:00:00.000000000 +0100
+++ linux-6.7-saam/include/dt-bindings/power/allwinner,sun8i-v853-ppu.h	2024-01-11 09:12:08.904246611 +0100
@@ -0,0 +1,10 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+
+#ifndef _DT_BINDINGS_POWER_SUN8I_V853_PPU_H_
+#define _DT_BINDINGS_POWER_SUN8I_V853_PPU_H_
+
+#define PD_RISCV		0
+#define PD_NPU			1
+#define PD_VE			2
+
+#endif
