diff -urN linux-6.7-rc4/drivers/rtc/rtc-sun6i.c linux-6.7-rc4-saam/drivers/rtc/rtc-sun6i.c
--- linux-6.7-rc4/drivers/rtc/rtc-sun6i.c	2023-12-03 10:52:56.000000000 +0100
+++ linux-6.7-rc4-saam/drivers/rtc/rtc-sun6i.c	2023-11-29 15:23:52.342481605 +0100
@@ -867,6 +867,8 @@
 		.data = (void *)RTC_LINEAR_DAY },
 	{ .compatible = "allwinner,sun50i-r329-rtc",
 		.data = (void *)RTC_LINEAR_DAY },
+	{ .compatible = "allwinner,sun8i-v853-rtc",
+		.data = (void *)RTC_LINEAR_DAY },
 	{ /* sentinel */ },
 };
 MODULE_DEVICE_TABLE(of, sun6i_rtc_dt_ids);
