diff -urN linux-6.7/drivers/gpu/drm/sun4i/sun4i_drv.c linux-6.7-saam/drivers/gpu/drm/sun4i/sun4i_drv.c
--- linux-6.7/drivers/gpu/drm/sun4i/sun4i_drv.c	2024-01-07 21:18:38.000000000 +0100
+++ linux-6.7-saam/drivers/gpu/drm/sun4i/sun4i_drv.c	2024-01-11 09:12:08.896246994 +0100
@@ -431,6 +431,7 @@
 	{ .compatible = "allwinner,sun8i-h3-display-engine" },
 	{ .compatible = "allwinner,sun8i-r40-display-engine" },
 	{ .compatible = "allwinner,sun8i-v3s-display-engine" },
+	{ .compatible = "allwinner,sun8i-v853-display-engine" },
 	{ .compatible = "allwinner,sun9i-a80-display-engine" },
 	{ .compatible = "allwinner,sun20i-d1-display-engine" },
 	{ .compatible = "allwinner,sun50i-a64-display-engine" },
diff -urN linux-6.7/drivers/gpu/drm/sun4i/sun8i_mixer.c linux-6.7-saam/drivers/gpu/drm/sun4i/sun8i_mixer.c
--- linux-6.7/drivers/gpu/drm/sun4i/sun8i_mixer.c	2024-01-07 21:18:38.000000000 +0100
+++ linux-6.7-saam/drivers/gpu/drm/sun4i/sun8i_mixer.c	2024-01-11 09:12:08.896246994 +0100
@@ -641,6 +641,15 @@
 	.vi_num		= 1,
 };
 
+static const struct sun8i_mixer_cfg sun8i_v853_mixer_cfg = {
+	.ccsc		= CCSC_MIXER0_LAYOUT,
+	.mod_rate	= 297000000,
+	.scaler_mask	= 0xf,
+	.scanline_yuv	= 2048,
+	.ui_num		= 1,
+	.vi_num		= 1,
+};
+
 static const struct sun8i_mixer_cfg sun50i_a64_mixer1_cfg = {
 	.ccsc		= CCSC_MIXER1_LAYOUT,
 	.mod_rate	= 297000000,
@@ -686,6 +695,10 @@
 		.data = &sun8i_v3s_mixer_cfg,
 	},
 	{
+		.compatible = "allwinner,sun8i-v853-de2-mixer",
+		.data = &sun8i_v853_mixer_cfg,
+	},
+	{
 		.compatible = "allwinner,sun20i-d1-de2-mixer-0",
 		.data = &sun20i_d1_mixer0_cfg,
 	},
